### PR TITLE
fix: pending tests, single/double quotes, logger

### PR DESCRIPTION
**Pull request by YaroSpace**
This is just to show the changes I have made locally.  Still investigating the json encoding of non-string objects.

### CODE REVIEW

File: lua/neotest-busted/output_handler.lua [hunk 1]
```diff
@@ -59,14 +63,23 @@ return function(options)
...
+            end,
+        })
 
         if ok then
             file:write(result)
             file:close()
         else
             io_write("Failed to encode test results to json: " .. result .. "\n")
             io_flush()
+            os.exit(1)
```
Comment (MisanthropicBit) [inline lua/neotest-busted/output_handler.lua:82]: Why is this necessary? Is it because the process cannot continue in a sane way if json encoding failed?

File: lua/neotest-busted/logging.lua [hunk 2]
```diff
@@ -16,6 +16,7 @@ local function log(level, context, message, ...)
     local args = { ... }
 
     vim.schedule(function()
+        if not unpack(args) then return end
```
Comment (MisanthropicBit) [inline lua/neotest-busted/logging.lua:19]: A better approach would be to just move assignment to `formatted_message` out of the callback to `vim.schedule` and make it `local formatted_message = message:format(…)`.

Comment (MisanthropicBit) [reply lua/neotest-busted/logging.lua:19]: #33

Comment (YaroSpace) [reply lua/neotest-busted/logging.lua:19]: 👍

File: lua/neotest-busted/util.lua [hunk 3]
```diff
@@ -6,7 +6,7 @@ local lib = require("neotest.lib")
 ---@param value string
 ---@return string
 function util.trim_quotes(value)
-    return vim.fn.trim(value, '"')
+    return vim.fn.trim(value, [["']])
```
[the remark this answers is not in the document]
Comment (YaroSpace) [reply lua/neotest-busted/util.lua:9]: 👍

File: lua/neotest-busted/output_handler.lua (hunk 1 above)
Comment (YaroSpace) [reply lua/neotest-busted/output_handler.lua:82]: Busted is often run as a separate process wrapped in `status, erro = pcal(.._l` for example. Setting the exit code, allows to assert the error.
I have actually updated the json formatter here https://github.com/lunarmodules/busted/pull/748/files

Comment (MisanthropicBit) [reply lua/neotest-busted/output_handler.lua:82]: I see but couldn't you assert on the error if you returned `false` as the second return value? Or called `error`? I don’t believe `os.exit` is necessary for `pcall`/`xpcall` to register an error.

Nice that you contributed to busted (and that it seems maintained). We can wait and see what the outcome of that PR is since it would probably be a good idea to align with what busted’s builtin handlers do.

File: lua/neotest-busted/init.lua [hunk 4]
```diff
@@ -380,6 +379,12 @@ function BustedNeotestAdapter.discover_positions(path)
         arguments: (arguments (_) @test.name (function_definition))
     ) (#match? @func_name "^it$")) @test.definition
 
+    ;; pending blocks
+    ((function_call
+        name: (identifier) @func_name
+        arguments: (arguments (_) @test.name (function_definition))
+    ) (#match? @func_name "^pending$")) @test.definition
+
```
Comment (MisanthropicBit) [inline lua/neotest-busted/init.lua:387]: I’ve had a chance to play around with this query and there are a few corner cases.

1. In the example you gave here https://github.com/MisanthropicBit/neotest-busted/issues/17#issuecomment-2495453094, the call to `pending` in the first test should not be matched, only the `it` test itself should. I think it needs match calls to `pending` whose parent is a `describe`. Running `busted —list <spec>` on the file shows what busted considers a test.
2. The query won’t match calls to `pending` without a second function argument.

I’ve got some working queries locally for these cases and I’ll continue to test it out a bit to make sure I haven’t missed something.

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: Totally agree.  That was a quick copy/paste from it block.

Comment (MisanthropicBit) [reply lua/neotest-busted/init.lua:387]: Fair enough, I figured it was. Here’s what I got so far. Haven’t found a better way yet to encode the parent-child relationship.

```scheme
;; pending blocks constrained to have a describe as a parent
((function_call
    name: (identifier) @func_name1 (#match? @func_name1 "^describe$")
    arguments: (arguments
        (_) @namespace.name
        (function_definition
        parameters: (parameters)
        body: (block
            ((function_call
                name: (identifier) @func_name2
                arguments: (arguments (_) @test.name (function_definition)?)
            ) (#match? @func_name2 "^pending$")
            ) @test.definition
        )
        )
    )
)) @namespace.definition
```

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: I am afraid this does not work properly with Neotest summary panel.
image
`it('this test does not run..)` should not come up in the panel, as it is inside the pending block'

I would also argue, that top level `pending` should be parsed as it does come up with `busted --list` and it is quite convenient to mark a whole buch of tests to pending by just renaming describe to pending.

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]:
```
    ;; pending blocks
    ((function_call
        name: (identifier) @func_name
        arguments: (arguments (string) @test.name)
    ) (#match? @func_name "^pending$")) @test.definition
```
this fixes it

Comment (MisanthropicBit) [reply lua/neotest-busted/init.lua:387]: Thanks for looking into this. I've been sick for the past 2 weeks and hadn't made much headway.

> it('this test does not run..) should not come up in the panel, as it is inside the pending block'

I can reproduce it. Not sure why exactly this happens.

> I would also argue, that top level pending should be parsed as it does come up with busted --list and it is quite convenient to mark a whole buch of tests to pending by just renaming describe to pending.

I agree. Constraining the pending to be inside a describe was a bit too restrictive.

I don't think we should constrain the argument to be a string since that would break parametric tests such as `it("a" .. i)`. The original query accounts for this.

> * in the case pending..it - all nested it are skipped anyway

In this case, the test fails and the output is something like `0 successes / 0 failures / 0 errors / 0 pending : 0.000567 seconds` since neotest-busted doesn't get any json output for the test. This may be acceptable for now.

> * in the case of it..pending, busted correctly marks the parent it as pending

I get the correct pending output for this case.

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: Sorry to hear about your illness.  I hope you are better now!

Ah, yes..forgot about parametric tests. You are right!

Comment (MisanthropicBit) [reply lua/neotest-busted/init.lua:387]: Thanks 🙂 it was just a bad case of the flu and I'm much better now.

If you have time you are welcome to try out #41 which I think can be merged as a good foundation for supporting pending tests.

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: I will have time this week, so sure, will do.

Comment (MisanthropicBit) [reply lua/neotest-busted/init.lua:387]: Thank you! Just FYI, I get a couple of false positives with the PR but I think it's still ok to merge for basic support.

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: Sorry for late reply, it was my time to catch a flu).
I have tested your changes on my set of tests and everything seems to be working just fine.

Comment (MisanthropicBit) [reply lua/neotest-busted/init.lua:387]: No problem, hope you get well soon. Thanks for taking the time to test things out!

Comment (YaroSpace) [reply lua/neotest-busted/init.lua:387]: If you run into any issues, I would be happy to help!
